plotly/tools.py

```diff
@@ -1451,6 +1455,823 @@ class FigureFactory(object):
+        Formats the 'median' hovertext for a violin plot.
+        """
+        from plotly.graph_objs import graph_objs
+
+        return graph_objs.Scatter(
+            x=[0],
+            y=[q2],
+            text=['median: ' + '{:0.2f}'.format(q2)],
+            mode='markers',
+            marker=dict(symbol='square',
+                        color='rgb(255,255,255)'),
+            hoverinfo='text'
+        )
+
+    @staticmethod
+    def _make_non_outlier_interval(d1, d2):
+        """
+        Returns the scatterplot fig of most of a violin plot.
+        """
+        from plotly.graph_objs import graph_objs
+
+        return graph_objs.Scatter(
+            x=[0, 0],
+            y=[d1, d2],
+            name='',
+            mode='lines',
+            line=graph_objs.Line(width=1.5,
+                                 color='rgb(0,0,0)')
+        )
+
+    @staticmethod
+    def _make_XAxis(xaxis_title, xaxis_range):
+        """
+        Makes the x-axis for a violin plot.
+        """
+        from plotly.graph_objs import graph_objs
+
+        xaxis = graph_objs.XAxis(title=xaxis_title,
+                                 range=xaxis_range,
+                                 showgrid=False,
+                                 zeroline=False,
+                                 showline=False,
+                                 mirror=False,
+                                 ticks='',
+                                 showticklabels=False,
+                                 )
+        return xaxis
+
+    @staticmethod
+    def _make_YAxis(yaxis_title):
+        """
+        Makes the y-axis for a violin plot.
+        """
+        from plotly.graph_objs import graph_objs
+
+        yaxis = graph_objs.YAxis(title=yaxis_title,
+                                 showticklabels=True,
+                                 autorange=True,
+                                 ticklen=4,
+                                 showline=True,
+                                 zeroline=False,
+                                 showgrid=False,
+                                 mirror=False)
+        return yaxis
+
+    @staticmethod
+    def _violinplot(vals, fillcolor=fillcolor, rugplot=True):
+        """
+        Refer to FigureFactory.create_violin() for docstring.
+        """
+        import numpy as np
+        from scipy import stats
+
+        vals = np.asarray(vals, np.float)
+        #  summary statistics
+        vals_min = FigureFactory._calc_stats(vals)['min']
+        vals_max = FigureFactory._calc_stats(vals)['max']
+        q1 = FigureFactory._calc_stats(vals)['q1']
+        q2 = FigureFactory._calc_stats(vals)['q2']
+        q3 = FigureFactory._calc_stats(vals)['q3']
+        d1 = FigureFactory._calc_stats(vals)['d1']
+        d2 = FigureFactory._calc_stats(vals)['d2']
+
+        # kernel density estimation of pdf
+        pdf = stats.gaussian_kde(vals)
+        # grid over the data interval
+        xx = np.linspace(vals_min, vals_max, 100)
```

Comment:
Why is this `xx` vs `x`? Is it supposed to point something out about the value the variable holds?